### PR TITLE
WT-12366 Fix the session prev prefetch ref flag usage to skip the page

### PR DESCRIPTION
**Pull request by kommiharibabu**
Modify how the ref parent page information is set for the current processing ref and compare it to the next ref parent to skip the page. The saved page pointer should never be accessed as it can be freed, it should only be used for the comparison.

### CODE REVIEW

File: src/btree/bt_prefetch.c [hunk 1]
```diff
@@ -25,10 +25,38 @@ __wt_btree_prefetch(WT_SESSION_IMPL *session, WT_REF *ref)
...
+     * evaluate to false and the counter will be reset, effectively marking the ref as available to
+     * pre-fetch from.
+     */
+    if (session->pf.prefetch_prev_ref_home == ref->home &&
+      session->pf.prefetch_skipped_with_parent < WT_PREFETCH_QUEUE_PER_TRIGGER) {
+        ++session->pf.prefetch_skipped_with_parent;
+        WT_STAT_CONN_INCR(session, block_prefetch_skipped_same_ref);
+        WT_STAT_CONN_INCR(session, block_prefetch_skipped);
+        return (0);
+    }
```
Comment (korteland) [inline src/btree/bt_prefetch.c:56]: What was the motivation behind moving this block?

File: src/btree/bt_prefetch.c [hunk 2]
```diff
@@ -48,18 +76,18 @@ __wt_btree_prefetch(WT_SESSION_IMPL *session, WT_REF *ref)
          * session transaction snapshot.
          */
         if (next_ref->state == WT_REF_DISK && F_ISSET(next_ref, WT_REF_FLAG_LEAF) &&
-          next_ref->page_del == NULL) {
+          next_ref->page_del == NULL && !F_ISSET(next_ref, WT_REF_FLAG_PREFETCH)) {
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_prefetch.c:79]: If we do not check for this `WT_REF_FLAG_PREFETCH` flag, we add the same `ref` again into the prefetch queue, which is not required.

File: src/btree/bt_prefetch.c (hunk 1 above)
Comment (kommiharibabu) [reply src/btree/bt_prefetch.c:56]: Earlier, this check was inside the function where we pushed the ref into the queue. All the child ref have the same parent page pointer. Checking it inside and ignoring it for all the child pages is just a waste of work. Instead, check it outside and return if the same parent page has already been processed.

File: src/btree/bt_prefetch.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (korteland) [reply src/btree/bt_prefetch.c:79]: Oh, of course. Sorry, not sure why I didn't spot that first time around. Thanks for explaining that.

File: src/btree/bt_prefetch.c (hunk 1 above)
Comment (korteland) [reply src/btree/bt_prefetch.c:56]: Got it - thank you.

**Review by korteland — APPROVED** [review]
LGTM! Thanks for looking into this.

File: src/btree/bt_prefetch.c [hunk 3]
```diff
@@ -25,10 +25,38 @@ __wt_btree_prefetch(WT_SESSION_IMPL *session, WT_REF *ref)
     conn = S2C(session);
     block_preload = 0;
 
-    if (!(F_ISSET(ref, WT_REF_FLAG_LEAF)) || (__wt_session_gen(session, WT_GEN_SPLIT) == 0))
-        return (WT_ERROR);
+    /* Don't add refs from trees that have eviction disabled since they are probably being closed */
```
Comment (clarissecheah) [inline src/btree/bt_prefetch.c:28]: NP: Could we add a fullstop to the end of this comment?

**Review by clarissecheah — APPROVED** [review]
LGTM % one nitpick. Thanks!

File: src/btree/bt_prefetch.c (hunk 3 above)
Comment (kommiharibabu) [reply src/btree/bt_prefetch.c]: Fixed.

**Pull request merged**